### PR TITLE
Code cleanup (replaces #8, #34, #39, #40, fixes #35, #38)

### LINKED ISSUE #35: Stripping newlines from the source tweets results in words being joined together
So you get messy tweets like [this one](https://twitter.com/00d4n1_ebooks/status/910399752336011264). I think instead of stripping newlines, all consecutive "runs" of whitespace (newlines, tabs, and spaces) should be collapsed to single spaces. Then the original meaning of the tweets is preserved and passed into the Markov generator, but spurious whitespace is avoided.

This can be achieved with something like
```python
tweet.text = re.sub('\s+', ' ', tweet.text)
```
I've made it an issue rather than jut doing it as a pull req since it might be worth discussing first.

### PR DESCRIPTION
**Pull request by ConorIA**
This pull request implements the changes suggested in #34, as well as some additional linting from python-flake8. 

This also includes the changes in #39 and #40, so merging this could close out all three of the other PRs. 

I probably should have waited for final feedback on #40, but I am happy to make changes to this PR based on the outcome of #40.

### CODE REVIEW

**Comment by tommeagher** [comment]
Hi @ConorIA. This all looks awesome. Thanks also for incorporating all the fixes @00dani had made. I'm inclined to merge this. The only thing missing I think is a little explanation of the new local_settings.
Would you be able to add a little bit to the "Configuring" section in the README? 
Otherwise, I think this is good to go.

**Comment by ConorIA** [comment]
Alright, done! I did make two small non-README changes. 

1. I set the `scrape` function to check the last URL, so that we only download a page once if a user wants to search it multiple times, as alluded to in the README this gives us less opportunities to hit an HTTP error or get flagged for hitting the same page multiple times. 
2. I added the space suggested in #8, which I had implemented in my production script, but had forgotten to copy back to my forked repo. Talk about an omnibus PR! 😜

File: ebooks.py [hunk 1]
```diff
@@ -34,119 +38,143 @@ def entity(text):
...
+
 def filter_tweet(tweet):
-    tweet.text = re.sub(r'\b(RT|MT) .+','',tweet.text) #take out anything after RT or MT
-    tweet.text = re.sub(r'(\#|@|(h\/t)|(http))\S+','',tweet.text) #Take out URLs, hashtags, hts, etc.
-    tweet.text = re.sub(r'\n','', tweet.text) #take out new lines.
-    tweet.text = re.sub(r'\"|\(|\)', '', tweet.text) #take out quotes.
-    tweet.text = re.sub(r'\s+\(?(via|says)\s@\w+\)?', '', tweet.text) # remove attribution
+    tweet.text = re.sub(r'\b(RT|MT) .+', '', tweet.text)  # take out anything after RT or MT
+    tweet.text = re.sub(r'(\#|@|(h\/t)|(http))\S+', '', tweet.text)  # Take out URLs, hashtags, hts, etc.
+    tweet.text = tweet.text.replace('\n', '')  # take out new lines.
```
Comment (tommeagher) [inline ebooks.py]: This should be 
```tweet.text = re.sub('\s+', ' ', tweet.text)```
per #35

File: markov.py [hunk 2]
```diff
@@ -68,7 +69,7 @@ def generate_sentence(self):
             sentence = ""
             for word in new_res:
                 sentence += word + " "
-            sentence += res[-2] + res[-1]
+            sentence += res[-2] + " " + res[-1]
```
Comment (tommeagher) [inline markov.py]: This adds an unnecessary space before the punctuation of every generated tweet. 
Like this: "Why can't I fix this via ." and "ALL THE THINGS AND WE WILL ."
This should be reverted to exclude this extra space.

File: ebooks.py [hunk 3]
```diff
@@ -34,119 +38,143 @@ def entity(text):
...
+            except Exception:
+                last_url = "ERROR"
+                import traceback
+                print(">>> Error scraping {0}:".format(src_url[i]))
+                print(traceback.format_exc())
+                continue
+            soup = BeautifulSoup(page, 'html.parser')
+        hits = soup.find_all(web_context[i], attrs=web_attributes[i])
+        for hit in hits:
+            tweet = str(hit.text).strip()
```
Comment (tommeagher) [inline ebooks.py]: This returns a UnicodeEncodeError in Python2. So there should be an if statement here to catch the exception.

File: local_settings_example.py [hunk 4]
```diff
@@ -2,17 +2,21 @@
...
 
-SOURCE_ACCOUNTS = [""] #A list of comma-separated, quote-enclosed Twitter handles of account that you'll generate tweets based on. It should look like ["account1", "account2"]. If you want just one account, no comma needed.
-ODDS = 8 #How often do you want this to run? 1/8 times?
-ORDER = 2 #how closely do you want this to hew to sensical? 2 is low and 4 is high.
-SOURCE_EXCLUDE = r'^$' #Source tweets that match this regexp will not be added to the Markov chain. You might want to filter out inappropriate words for example.
-DEBUG = True #Set this to False to start Tweeting live
-STATIC_TEST = False #Set this to True if you want to test Markov generation from a static file instead of the API.
-TEST_SOURCE = ".txt" #The name of a text file of a string-ified list for testing. To avoid unnecessarily hitting Twitter API. You can use the included testcorpus.txt, if needed.
-TWEET_ACCOUNT = "" #The name of the account you're tweeting to.
+SOURCE_ACCOUNTS = [""]  # A list of comma-separated, quote-enclosed Twitter handles of account that you'll generate tweets based on. It should look like ["account1", "account2"]. If you want just one account, no comma needed.
```
Comment (tommeagher) [inline local_settings_example.py]: We should mention that you *must* include at least one twitter account here, or the script fails with an ```IndexError``` on line 139 of ebooks.py

File: ebooks.py (hunk 1 above)
Comment (ConorIA) [reply ebooks.py]: Done.

File: ebooks.py [hunk 5]
```diff
@@ -34,119 +38,143 @@ def entity(text):
...
+    tweet.text = re.sub(r'\"|\(|\)', '', tweet.text)  # take out quotes.
+    tweet.text = re.sub(r'\s+\(?(via|says)\s@\w+\)?', '', tweet.text)  # remove attribution
     htmlsents = re.findall(r'&\w+;', tweet.text)
-    if len(htmlsents) > 0 :
-        for item in htmlsents:
-            tweet.text = re.sub(item, entity(item), tweet.text)    
-    tweet.text = re.sub(r'\xe9', 'e', tweet.text) #take out accented e
+    for item in htmlsents:
+        tweet.text = tweet.text.replace(item, entity(item))
+    tweet.text = tweet.text.replace('\xe9', 'e')  # take out accented e
```
[the remark this answers is not in the document]
Comment (ConorIA) [reply ebooks.py]: Insofar as the change was to make the code "cleaner", I think a try-except block is considerable "messier". Reverted.

File: ebooks.py (hunk 3 above)
Comment (ConorIA) [reply ebooks.py]: Handled! I think the `UnicodeDecodeError` might be superfluous. Let me know if you'd like me to remove.

File: local_settings_example.py (hunk 4 above)
Comment (ConorIA) [reply local_settings_example.py]: Handled in code by first testing to make sure that `SOURCE_ACCOUNTS` is not empty.

File: markov.py (hunk 2 above)
Comment (ConorIA) [reply markov.py]: Fixed. Please test. I'm not sure how Pythonic the fix is, but it seems to work!

Comment (tommeagher) [reply markov.py]: Not very Pythonic, but it seems to work fine, so it's good by me. Thank you!

**Comment by tommeagher** [comment]
This all looks swell. Thanks, @ConorIA for everything. Merging now.

**Pull request merged**